xds/src/main/java/io/grpc/xds/EdsLoadBalancer2.java

```diff
@@ -105,7 +105,9 @@ public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     if (cluster == null) {
       cluster = config.clusterName;
     }
-    switchingLoadBalancer.switchTo(new EdsLbState(config.edsServiceName, config.lrsServerName));
+    switchingLoadBalancer.switchTo(new EdsLbState(
+        config.edsServiceName, config.lrsServerName, config.localityPickingPolicy,
```

Comment:
This looks more than a fix. You changed the definition of `EdsLbState`, why? `But `EdsLbState` `equals()` and `hashCode()` are not changed accordingly.

Reply:
That's done purposely. `EdsLbState` is a `LoadBalancer.Factory` used for switching in eds_service_name or lrs_server_name change (aka, replacing cluster). But we don't treat changing endpoint_picking_policy or locality_picking_policy as replacing cluster. So they are not part of equality of `EdsLbState`. These two fields are mutable and are moved to `EdsLbState` from `ChildLbState` to avoid the watcher reentrancy issue: `ChildLbState` is subscribing to some endpoint resource, while the addWatcher operation can invoke the callback inline (reentrant). So it is too late to set up ChildLbState's  endpoint_picking_policy and locality_picking_policy in its handleResolvedAddresses.

Reply:
I'm not so comfortable having these mutable fields in a Factory. It looks not elegant. Is it possible to move `xdsClient.watchEndpointData()` in `handleResolvedAddresses()` to avoid reentrancy issue?

Reply:
I tried a couple of different approaches, none of them seems to be satisfactory. `handleResolvedAddresses` is called for multiple times while `xdsClient.watchEndpointData()` is only called once. Moving it into `handleResolvedAddresses` is cumbersome. From another aspect, `EdsLbState` is not just a LoadBalancer.Factory, it implements `LoadBalancer.Factory` because we have to (because of the usage of GracefulSwitchLoadBalancer API), but it's also representing the state of `EdsLoadBalancer`: a running instance for one EdsConfig (eds_service_name, lrs_server_name, locality_picking_policy, endpoint_picking_policy). From this perspective, it's not that weird to have endpoint_picking_policy and locality_picking_policy saved there.

Reply:
Moved `endpoint_picking_policy` and `locality_picking_policy` into `ChildLbState`. Delaying the execution of `xdsClient.watchEndpointData()` outside `ChildLbState`'s constructor to avoid reentrancy may be problematic, as `ChildLbState.shutdown()` (which cancels the watcher) may happen before current task in the syncContext returns. So it'd be better to keep watcher addition/cancellation in place. Instead, we can just let `onEndpointChanged` callback not invoking child policy's `handleResolvedAddresses()` if such a reentrancy happens. After this callback returns, the following invocation of `ChildLbState`'s `handleResolvedAddresses()` will get the endpoint_picking_policy and locality_picking_policy and it will invoke child policy's `handleResolvedAddresses()`.